pkg/server/server.go

```diff
@@ -283,6 +283,7 @@ func NewServerWithOpts(opts ...OpenFGAServiceV1Option) (*Server, error) {
 	s.checkOptions = []graph.LocalCheckerOption{
 		graph.WithResolveNodeBreadthLimit(s.resolveNodeBreadthLimit),
 		graph.WithMaxConcurrentReads(s.maxConcurrentReadsForCheck),
+		graph.WithSingleflightResolver(),
```

Comment:
If in the future we need to create a new `LocalChecker` somewhere else in the codebase, we will need to remember to add this option in, or we will lose out on the benefits of this PR.

This is why i don't like having this as an option 😄

Reply:
I agree. Now that we have a few layers in the "onion", this way of configuring the resolver with the cache and singleflight resolver are prone to omission or incorrect ordering. I'm not convinced that its time to refactor this configuration, but definitely the next time a nested resolver is created.

Reply:
**Update:** I agree in principle that singleflight should be opt-in rather than opt-out but a couple issues I faced when trying to implement that:

- It is inconsistent with the opt-in mechanism of caching. I would argue that both are important and both should be opt-out.
- Ordering of resolvers is important. The singleflight resolver needs to be added before the cache resolver. If we introduce a `WithNoSingleflightResolver` we'll need to evaluate these options very carefully and ensure they're ordered correctly to be compliant with the current way of doing things.

If we want these options to be opt-out, we should refactor the existing method of passing-in resolver options to be more consistent across all resolvers. I don't think it would be appropriate at this time to introduce a

Finally, this only affects instantiation of a new local checker. When instantiating a server, singleflight is not something you can even opt-out of at this time. As far as I can tell, we instantiate new local checkers for testing specific scenarios and perhaps its better to have the current level of flexibility, even if there is a tradeoff with needing to remember to add the singleflight in some cases.

Reply:
My two-cents : If the opt-out intent is largely focused around simplification and avoidance of forgetting the SingleflightCheckResolver, then that is and should be addressed with better tests around the consumer code that expects that composition to have been setup properly. Presumably, if you are testing for de-dupe behavior or performance you should cover your bases. If not, then the tests are poor, not the composition structure.

I wouldn't encourage us to make the decision around the test code so much as I would around the compositional intent. The intent of the delegate and CheckResolver pattern is iterative wrapping, and wrapping should be an opt-in feature. A consumer lib should choose to wrap if their intent is to have SingleflightCheckResolver behavior.

Reply:
In fact, this composition is currently broken (both before this PR and in this PR). The server should construct a `SingleflightCheckResolver`, a `CachedCheckResolver` and a `LocalChecker` and set their delegates up appropriately and make sure the first resolver is the top-most one, respectively. A naive, but demonstrative example of what I'm referring to would look like:

```go
// pkg/server/server.go

func Check(ctx context.Context, req *openfgav1.CheckRequest) (*openfgav1.CheckResponse, error) {
    singleflightChecker := NewSingleflightCheckResolver(...)
    cachedChecker := NewCachedCheckResolver(...)
    localChecker := NewLocalChecker(...)
    
    singleflightChecker.SetDelegate(cachedChecker)
    cachedChecker.SetDelegate(localChecker)
    localChecker.SetDelegate(singleflightChecker)

    // notice usage of __singleflightChecker__ here
    resolveCheckResp, err := singleflightChecker.ResolveCheck(ctx, ...)
    
    resp := transform(resolveCheckResp) // transform this to *openfgav1.CheckResponse

   return resp, err
}
```

Today we're actually delegating first to the `LocalChecker#ResolveCheck`, and not the `CachedCheckResolver#ResolveCheck`, and in this PR we're delegating to `LocalCheck#ResolveCheck` first still as well. That needs to change.